query.go

```diff
@@ -530,7 +529,7 @@ func (dht *IpfsDHT) dialPeer(ctx context.Context, p peer.ID) error {
 	defer span.End()
 
 	// short-circuit if we're already connected.
-	if dht.host.Network().Connectedness(p) == network.Connected {
+	if HasValidConnectedness(dht.host, p) {
```

Comment:
We need a non-limited connection here.

Reply:
In other words, if the connected is limited, and we don't short-circuit here, `dht.host.Connect` won't necessarily hole-punch. 

Am I missing anything?

Reply:
@2color you're right. This is a bug in go-libp2p. See: https://github.com/libp2p/go-libp2p/issues/2714

I think the best you can do right now is poll for a few seconds, if the connection has been upgraded to a NonLimited Connection.

Reply:
@sukunrt Under what circumstances does go-libp2p attempt to hole punch? Does it do this automatically after establishing a circuit relay reservation?

Reply:
> Does it do this automatically after establishing a circuit relay reservation?

Yes. The receiver initiates the hole punch. If you're the one dialing the p2p-circuit address, the peer is the one who will initiate the hole punch. So all we can do is just wait for the peer to initiate the hole punch.